usermods/ULC_Battery_Management/usermod_ulc_batterymanagement.h

```diff
@@ -0,0 +1,279 @@
+#pragma once
+
+#include "wled.h"
+#include "IP5306_I2C.h"
+#include <driver/adc.h>
+
+// the frequency to check battery voltage, 5 seconds
+#ifndef USERMOD_BATTERY_MEASUREMENT_INTERVAL
+  #define USERMOD_BATTERY_MEASUREMENT_INTERVAL 5000
+#endif
+
+#ifndef USERMOD_BATTERY_ADC_PRECISION
+  #define USERMOD_BATTERY_ADC_PRECISION 4095.0f
+#endif
+
+#ifndef USERMOD_BATTERY_ADC_REF
+  #define USERMOD_BATTERY_ADC_REF 3.3f
+#endif
+
+#ifndef USERMOD_BATTERY_ADC_R1
+  #define USERMOD_BATTERY_ADC_R1 8.0f
+#endif
+
+#ifndef USERMOD_BATTERY_ADC_R2
+  #define USERMOD_BATTERY_ADC_R2 2.5f
+#endif
+
+#ifndef USERMOD_BATTERY_MIN_VOLTAGE
+  #define USERMOD_BATTERY_MIN_VOLTAGE 2.6f
+#endif
+
+#ifndef USERMOD_BATTERY_MAX_VOLTAGE
+  #define USERMOD_BATTERY_MAX_VOLTAGE 4.2f
+#endif
+
+#ifndef USERMOD_IP5306_SDA
+  #define USERMOD_IP5306_SDA 32
+#endif
+
+#ifndef USERMOD_IP5306_SCL
+  #define USERMOD_IP5306_SCL 33
+#endif
+
+#ifndef USERMOD_ABL_BATTERY
+  #define USERMOD_ABL_BATTERY 1250
+#endif
+
+#ifndef USERMOD_ABL_EXT
+  #define USERMOD_ABL_EXT 2500
+#endif
+
+class UsermodULCBatteryManagement : public Usermod {
+
+  private:
+    IP5306 *ip5306 = nullptr;
+
+    bool initDone = false;
+    bool initializing = true;
+    // GPIO pin used for battery measurment (with a default compile-time fallback)
+    int8_t ip5306_sda = USERMOD_IP5306_SDA;
+    int8_t ip5306_scl = USERMOD_IP5306_SCL;
+    // how often do we measure?
+    unsigned long readingInterval = USERMOD_BATTERY_MEASUREMENT_INTERVAL;
+    unsigned long nextReadTime = 0;
+    unsigned long lastReadTime = 0;
+
+    uint32_t ablBattery = USERMOD_ABL_BATTERY;
+    uint32_t ablExt = USERMOD_ABL_EXT;
+    
+    // battery min. voltage
+    float minBatteryVoltage = USERMOD_BATTERY_MIN_VOLTAGE;
+    // battery max. voltage
+    float maxBatteryVoltage = USERMOD_BATTERY_MAX_VOLTAGE;
+
+    uint8_t batteryLevel = 0; 
+    bool isCharging = false;
+    bool isPluggedIn = false;
+
+    // strings to reduce flash memory usage (used more than twice)
+    static const char _name[];
+    static const char _readInterval[];
+    static const char _maxCurrentBattery[];
+    static const char _maxCurrentExt[];
+
+    // custom map function
+    // https://forum.arduino.cc/t/floating-point-using-map-function/348113/2
+    double mapf(double x, double in_min, double in_max, double out_min, double out_max) 
+    {
+      return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
+    }
+
+    float truncate(float val, byte dec) 
```

Comment:
this method to "truncate" a float to a number of decimals is _horribly_ inefficient due to the use of `pow()`. Please try to replace it.

Reply:
I seem to have forgotten to remove this function as I do not need it anymore, anyways. I have removed it now